### PR TITLE
Added bounding box paste for CPU backend.

### PR DESCRIPTION
**Pull request by mzient**
BBoxPaste operator resizes and moves bounding boxes so that they describe appropriate regions of the original image after it's pasted over a larger canvas.
Relevant parameters are the same as in ops.Paste.
ltrb parameter switches between [x, y, width, height] (the default) and [left, top, right, bottom] representations of bounding boxes.

### CODE REVIEW

File: dali/pipeline/operators/paste/bbox_paste.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+#include "dali/common.h"
+#include "dali/error_handling.h"
+#include "dali/pipeline/operators/common.h"
+#include "dali/pipeline/operators/operator.h"
+
+namespace dali {
+
+template <typename Backend>
+class BBoxPaste : public Operator<Backend> {
```
Comment (Kh4L) [inline dali/pipeline/operators/paste/bbox_paste.h]: You could  specialize it to CPUBackend directly here, since the run implem supports only a CPUBackend.

Comment (mzient) [reply dali/pipeline/operators/paste/bbox_paste.h]: Given the recent trend of adding GPU versions of bounding box processing stages, it seemed reasonable to keep it generic for future GPU variant.

Comment (Kh4L) [reply dali/pipeline/operators/paste/bbox_paste.h]: I thought that we had another implem for the GPU version
So you can template specialize RunImpl for CPUBackend here

Comment (mzient) [reply dali/pipeline/operators/paste/bbox_paste.h]: Specialization of the entire class is ok. What I don't want here is something like
`class BbFlip : public Operator<CPUBackend>`
Instead, we'd have

`template<>
class BBoxPaste<CPUBackend> : public Operator<CPUBackend>`

Is that what you had in mind?

Comment (Kh4L) [reply dali/pipeline/operators/paste/bbox_paste.h]: I was thinking about something similar as what we do here, specialize the implem in the .cc :
https://github.com/NVIDIA/DALI/blob/9ec7d0af11d1d207eee722a81c50f869a91f5a81/dali/pipeline/operators/crop/slice.cc#L59

File: dali/pipeline/operators/paste/bbox_paste.cc [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    x0 = x0 * scale + ofsx;
+    y0 = y0 * scale + ofsy;
+    if (useLTRB) {
+      x1w = x1w * scale + ofsx;
+      y1h = y1h * scale + ofsy;
+    } else {
+      x1w = x1w * scale;
+      y1h = y1h * scale;
+    }
+    // CAVEAT: Precision issues may drive bounding box outside (0-1).
```
Comment (pribalta) [inline dali/pipeline/operators/paste/bbox_paste.cc]: If this happens, RandomBBoxCrop will crash. Would be good to align on that and decide what is our assumption about bounding boxes. If we only support [0,1] (like atm) these values have to be clamped

File: dali/pipeline/operators/paste/bbox_paste.h [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+ public:
+  explicit inline BBoxPaste(const OpSpec &spec) :
+    Operator<Backend>(spec) {
+    useLTRB = spec.GetArgument<bool>("ltrb");
+  }
+
+ protected:
+  bool useLTRB = false;
```
Comment (pribalta) [inline dali/pipeline/operators/paste/bbox_paste.h]: very minor but according to our coding standard it should be useLTRB_

File: dali/pipeline/operators/paste/bbox_paste.cc (hunk 2 above)
Comment (mzient) [reply dali/pipeline/operators/paste/bbox_paste.cc]: I need to dig into it a bit, but it seems it's not a matter of simple clamping (with LTRB it is, but not with width/height). My educated guess is that it'd require some nextafter calls to be sure that nothing breaks.

File: dali/pipeline/operators/paste/bbox_paste.h (hunk 3 above)
Comment (szalpal) [reply dali/pipeline/operators/paste/bbox_paste.h]: Even further, `useLtrb_` ;)

After https://google.github.io/styleguide/cppguide.html#General_Naming_Rules:

> For some symbols, this style guide recommends names to start with a capital letter and to have a capital letter for each new word (a.k.a. "Camel Case" or "Pascal case"). When abbreviations or acronyms appear in such names, prefer to capitalize the abbreviations or acronyms as single words (i.e StartRpc(), not StartRPC()).

File: dali/pipeline/operators/paste/bbox_paste_test.cc [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+    ToTensorList(*out, boxes);
+    return out;
+  }
+
+
+  vector<TensorList<CPUBackend>*>
+  Reference(const vector<TensorList<CPUBackend>*> &inputs,
+            DeviceWorkspace *ws)
+  {
+    (void)inputs;
```
Comment (JanuszL) [inline dali/pipeline/operators/paste/bbox_paste_test.cc]: Maybe just remove this variable names from the argument list.

File: docs/examples/detection_pipeline.ipynb [hunk 5]
```diff
@@ -13,7 +13,13 @@
    "cell_type": "code",
    "execution_count": 1,
    "metadata": {},
-   "outputs": [],
+   "outputs": [
```
Comment (JanuszL) [inline docs/examples/detection_pipeline.ipynb]: Remove

File: docs/examples/detection_pipeline.ipynb [hunk 6]
```diff
@@ -98,13 +122,20 @@
   },
   {
    "cell_type": "code",
-   "execution_count": 1,
+   "execution_count": null,
```
Comment (JanuszL) [inline docs/examples/detection_pipeline.ipynb]: Remove

**Comment by JanuszL** [comment]
I see that juyter is broken now -  or at least I cannot display it.

**Comment by JanuszL** [comment]
Builds 39599

**Pull request merged**